### DISCUSSION TITLE
diff:Driver alternatives

### DISCUSSION THREAD

**Discussion by avivanoff** [wixtoolset/issues#7571] [Questions]
When using WixToolset.DifxApp Driver element, we get the following warning:

`WIX1130: The Driver has been deprecated.`

What are non-deprecated alternatives?

**Comment by BMurri** [comment]
The Microsoft docs say to use SetupApi. I don't know of any open source implementation of a WiX extension that implements this.

Contributions would certainly be welcome.

**Reply by oleg-shilo** [reply]
> MyProduct.wxs(13) : warning WIX1130: The Driver element has been deprecated.

Maybe then it makes sense to requalify the warning to the error?

**Reply by barnson** [reply]
deprecated != obsoleted

**Reply by oleg-shilo** [reply]
I agree but the on a practical level with WiX4 `obsoleted==error` as the use of `dixf:Driver` leads to the internal error. Both the warning and the error are triggered by the same line:

Input:
```xml
<Wix xmlns="http://wixtoolset.org/schemas/v4/wxs" xmlns:difx="http://wixtoolset.org/schemas/v4/wxs/difxapp">
  <Package Compressed="yes" ProductCode="6f330b47-2577-43ad-9095-1861ca25889c" Name="MyProduct" Language="1033" Version="1.0.0.0" UpgradeCode="6f330b47-2577-43ad-9095-1861ba25889b" Manufacturer="oleg.shilo" InstallerVersion="200">
    <SummaryInformation Codepage="Windows-1252" />
    <Media Id="1" Cabinet="MyProduct.cab" EmbedCab="yes" />

    <StandardDirectory Id="ProgramFilesFolder">
      <Directory Id="My20Company" Name="My Company">
        <Directory Id="INSTALLDIR" Name="My Device">

          <Component Id="Component.driver.sys_2218891202" Guid="6f330b47-2577-43ad-9095-1861767c3cdf">
            <File Id="driver.sys_2218891202" Source="driver.sys" />

            <difx:Driver AddRemovePrograms="no" DeleteFiles="no" Legacy="yes" PlugAndPlayPrompt="no" Sequence="1" />
          </Component>

        </Directory>
      </Directory>
    </StandardDirectory>

    <WixUI Id="WixUI_Minimal" xmlns="http://wixtoolset.org/schemas/v4/wxs/ui" />

    <UIRef Id="WixUI_ErrorProgressText" />

    <Feature Id="Complete" Title="Complete" Level="1" AllowAbsent="yes">
      <ComponentRef Id="Component.driver.sys_2218891202" />
    </Feature>
  </Package>
</Wix>
```

```txt
wix build  -ext WixToolset.DifxApp.wixext -ext WixToolset.UI.wixext MyProduct.wxs  -o MyProduct.msi
```
Output:

```txt
MyProduct.wxs(13) : warning WIX1130: The Driver element has been deprecated.
MyProduct.wxs(13) : error WIX0094: The identifier 'CustomAction:MsiProcessDrivers' could not be found. Ensure you have typed the reference correctly and that all the necessary inputs are provided to the linker.
```
I will be very happy if I ma wrong and there is a mistake in my code. 
I have attached the sample.

[Driver.zip](https://github.com/wixtoolset/issues/files/12061056/Driver.zip)

**Reply by barnson** [reply]
https://wixtoolset.org/docs/schema/difxapp/driver/

**Reply by oleg-shilo** [reply]
What exactly are you trying to say by providing this link to the documentation?
Did you see the mistake in the code I used? Inconsistency with that documentation?
Did the documentation describe the error the compiler generates?

I am not sure how to interpret your single-line response. :(

**Reply by BMurri** [reply]
> Installs a driver. To use this element, you need to reference the WixDifxAppExtension extension and add the .wixlib appropriate for the target platform (difxapp_x86.wixlib or difxapp_x64.wixlib) to your project.

From the linked page. I'll bold it for emphasis.

**To use this element, you need to reference the WixDifxAppExtension extension and add the .wixlib appropriate for the target platform (difxapp_x86.wixlib or difxapp_x64.wixlib) to your project.**

**Reply by BMurri** [reply]
BTW, the warning and the error are unrelated.

**Reply by oleg-shilo** [reply]
Excellent. Thank you

For everyone who is to use reads this. The problem was the absence of the `-lib` argument:

```txt
wix build  -ext WixToolset.DifxApp.wixext -ext WixToolset.UI.wixext -lib %userprofile%\.wix\extensions\WixToolset.DifxApp.wixext\4.0.0\wixext4\difxapp_x64.wixlib" MyProduct.wxs  -o MyProduct.msi
```

**Comment by avivanoff** [comment]
I have [another problem](https://github.com/wixtoolset/issues/issues/7625) using WixToolset.DifxApp.wixext.

**Comment by oleg-shilo** [comment]
After the discussion in the https://github.com/wixtoolset/issues/issues/7625 thread WiX team member confirmed that currently there is an unresolved issue with the `Driver` element.

**Reply by robmen** [reply]
Only if you use the Driver extension in a .wixlib.

**Reply by oleg-shilo** [reply]
Rob, can you elaborate on "Only if ...".

From the link you provided, I take that the only way to use Driver elements is to add `.wixlib`:

> To use this element, you need to reference the WixDifxAppExtension extension **and add the .wixlib** appropriate for the target platform (difxapp_x86.wixlib or difxapp_x64.wixlib) to your project.

Does your last post mean that it is possible to use Driver element without `.wixlib`?

**Reply by BMurri** [reply]
> Rob, can you elaborate on "Only if ...".
> 
...
> 
> Does your last post mean that it is possible to use Driver element without `.wixlib`?

There's "with" and there's "in". The build context matters.

There are no known issues (apart from the depreciation of DIFxApp) with using that element when building a package. If building using the `wix` tool, you might need to also include the extension's wixlib (when using msbuild the extension's wixlib is included by `PackageReference`).

There is a now known issue with using that element when building a wixlib. That's what Rob is calling out.

**Reply by oleg-shilo** [reply]
Thank you Blair,

Then I will finalize the findings of this thread. Basically what you said.

- When compiling/building wxs containing `Driver` element with `wix.exe` both `WixToolset.DifxApp.wixext` and `difxapp_x??.wixlib` need to be passed to the tool as command line args

- When compiling/building wxs containing `Driver` element with `msbuild.exe` (or Visual Studio) only `WixToolset.DifxApp.wixext` needs to be referenced.
